File: src/mvmctl/core/_shared/_guestfs/_service.py
from __future__ import annotations

import logging
import os
import shutil
import subprocess
from pathlib import Path

from mvmctl.utils._system import ProcessSignalHandler, has_python_ancestor
from mvmctl.utils.common import CacheUtils

from ._kernel_detector import KernelDetector

logger = logging.getLogger(__name__)


class GuestfsService:
    """Stateless service for libguestfs appliance and backend operations."""

    @staticmethod
    def build_appliance(cache_dir: Path) -> Path | None:
        """Build the libguestfs fixed appliance for faster image operations.

        Uses KernelDetector to find a suitable upstream kernel with virtio
        drivers, sets the appropriate environment variables, and runs
        libguestfs-make-fixed-appliance.

        Args:
            cache_dir: Base cache directory where the appliance will be built.

        Returns:
            Path to appliance directory if built, None if skipped or failed.
        """
        make_tool = shutil.which("libguestfs-make-fixed-appliance")
        if not make_tool:
            logger.debug(
                "libguestfs-make-fixed-appliance not found — skipping appliance build"
            )
            return None

        appliance_dir = cache_dir / "appliance"
        appliance_dir.mkdir(parents=True, exist_ok=True)

        required_files = {"kernel", "initrd", "root"}
        if required_files.issubset({p.name for p in appliance_dir.iterdir()}):
            logger.debug(
                "libguestfs appliance already present at %s", appliance_dir
            )
            return appliance_dir

        GuestfsService.clean_stale_guestfs_state()

        env = os.environ.copy()
        kernel_info = KernelDetector.find_best_kernel()
        if kernel_info is not None:
            kernel_path, modules_dir = kernel_info
            env["SUPERMIN_KERNEL"] = str(kernel_path)
            env["SUPERMIN_MODULES"] = str(modules_dir)
            logger.debug(
                "Forcing libguestfs appliance build with kernel %s",
                kernel_path,
            )
        else:
            logger.warning(
                "No suitable kernel with virtio drivers found in /boot — "
                "appliance build may hang if the auto-selected kernel lacks virtio"
            )

        try:
            subprocess.run(
                [make_tool, str(appliance_dir)],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.PIPE,
                text=True,
                check=True,
                timeout=60,
                env=env,
            )
        except subprocess.TimeoutExpired:
            logger.warning("libguestfs appliance build timed out after 60s")
            return None
        except subprocess.CalledProcessError as e:
            logger.warning("libguestfs appliance build failed: %s", e.stderr)
            return None
        except FileNotFoundError:
            return None
        else:
            logger.info("libguestfs fixed appliance built at %s", appliance_dir)
            return appliance_dir

    @staticmethod
    def clean_stale_guestfs_state() -> bool:
        """Remove stale libguestfs processes, locks, sockets, and caches.

        First, kills any abandoned QEMU/guestfish processes that are running
        the libguestfs appliance but have no active Python/mvmctl ancestor.
        Then cleans up lock files, daemon sockets, and cached appliance
        directories that can cause subsequent appliance operations to hang.

        Returns:
            True if any stale state was removed or process was killed.
        """
        uid = os.getuid()
        cleaned = False

        # ── Phase 0: Find and kill abandoned guestfs processes ──────────
        abandoned_pids = GuestfsService._find_abandoned_guestfs_processes(uid)
        if abandoned_pids:
            ProcessSignalHandler.terminate_batch(
                abandoned_pids,
                graceful_timeout=0.5,
            )
            cleaned = True

        # ── Phase 1: Remove the global lock file ────────────────────────
        lock_file = Path(f"/var/tmp/.guestfs-{uid}/lock")
        if lock_file.exists():
            try:
                lock_file.unlink()
                cleaned = True
                logger.debug("Removed stale libguestfs lock: %s", lock_file)
            except OSError:
                pass

        # ── Phase 2: Remove stale daemon sockets ────────────────────────
        for sock_dir in Path(f"/run/user/{uid}").glob("libguestfs*"):
            for sock in sock_dir.glob("guestfsd.sock"):
                try:
                    sock.unlink()
                    cleaned = True
                    logger.debug("Removed stale libguestfs socket: %s", sock)
                except OSError:
                    pass

        # ── Phase 3: Remove cached appliance directories in /var/tmp ────
        guestfs_tmp = Path(f"/var/tmp/.guestfs-{uid}")
        if guestfs_tmp.exists():
            for entry in guestfs_tmp.glob("appliance.d*"):
                if entry.is_dir():
                    try:
                        shutil.rmtree(entry)
                        cleaned = True
                        logger.debug(
                            "Removed stale libguestfs cache: %s", entry
                        )
                    except OSError:
                        pass

        return cleaned

    @staticmethod
    def _find_abandoned_guestfs_processes(uid: int) -> list[int]:
        """Find QEMU/guestfish PIDs owned by *uid* that are running the
        libguestfs appliance but have no Python/mvmctl ancestor.

        These are orphaned processes left behind after crashes or failed
        appliance builds.
        """
        abandoned: list[int] = []

        proc_path = Path("/proc")
        if not proc_path.exists():
            return abandoned

        try:
            for proc_dir in proc_path.iterdir():
                if not proc_dir.name.isdigit():
                    continue
                pid = int(proc_dir.name)
                try:
                    proc_uid = proc_dir.stat().st_uid
                except OSError:
                    continue
                if proc_uid != uid:
                    continue

                # Read cmdline and check for guestfs signature
                try:
                    cmdline_raw = (proc_dir / "cmdline").read_bytes()
                    cmdline = cmdline_raw.decode("utf-8", errors="replace")
                except OSError:
                    continue

                # Match QEMU processes running the libguestfs appliance,
                # or guestfish processes left behind.
                if (
                    ".guestfs-" in cmdline
                    and ("appliance.d" in cmdline or "guestfsd.sock" in cmdline)
                ) or "guestfish" in cmdline.lower():
                    if not has_python_ancestor(pid):
                        abandoned.append(pid)
        except FileNotFoundError:
            pass

        return abandoned

    @staticmethod
    def prune_appliance(dry_run: bool = False) -> bool:
        """Remove the libguestfs appliance folder and stale system state.

        Also cleans up stale locks and sockets in /var/tmp and /run/user
        that can cause subsequent appliance builds to hang.

        Args:
            dry_run: If True, only report what would be removed.

        Returns:
            True if appliance folder or stale state was removed.
        """
        appliance_dir = CacheUtils.get_cache_dir() / "appliance"
        removed = False
        if appliance_dir.exists():
            if not dry_run:
                shutil.rmtree(appliance_dir, ignore_errors=True)
            removed = True

        if not dry_run:
            state_cleaned = GuestfsService.clean_stale_guestfs_state()
            removed = removed or state_cleaned

        return removed

File: internal/infra/guestfs/service.go
package guestfs

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
	"time"

	"mvmctl/internal/infra/system"
)

// ── GuestfsService ──────────────────────────────────────────────────────────
//
// Mirrors src/mvmctl/core/_shared/_guestfs/_service.py GuestfsService.
// Stateless service for libguestfs appliance and backend operations.

// GuestfsService provides static helpers for libguestfs appliance management.
type GuestfsService struct{}

// BuildAppliance builds the libguestfs fixed appliance for faster image ops.
// Uses KernelDetector to find a suitable upstream kernel with virtio drivers,
// sets the appropriate environment variables, and runs
// libguestfs-make-fixed-appliance.
//
// Returns the path to the appliance directory if built, or empty string if
// skipped or failed.
func (gs *GuestfsService) BuildAppliance(ctx context.Context, cacheDir string) (string, error) {
	makeTool, err := exec.LookPath("libguestfs-make-fixed-appliance")
	if err != nil {
		slog.Debug("libguestfs-make-fixed-appliance not found — skipping appliance build")
		return "", nil
	}

	applianceDir := filepath.Join(cacheDir, "appliance")
	if err := os.MkdirAll(applianceDir, 0755); err != nil {
		return "", fmt.Errorf("create appliance dir: %w", err)
	}

	// Check if appliance already exists (has kernel, initrd, root files)
	requiredFiles := map[string]bool{"kernel": false, "initrd": false, "root": false}
	entries, err := os.ReadDir(applianceDir)
	if err == nil {
		for _, e := range entries {
			if _, ok := requiredFiles[e.Name()]; ok {
				requiredFiles[e.Name()] = true
			}
		}
		allPresent := true
		for _, present := range requiredFiles {
			if !present {
				allPresent = false
				break
			}
		}
		if allPresent {
			slog.Debug("libguestfs appliance already present", "path", applianceDir)
			return applianceDir, nil
		}
	}

	// Clean stale state first
	gs.CleanStaleGuestfsState()

	// Build environment
	env := os.Environ()
	kd := &KernelDetector{}
	kernelPath, modulesDir, kerr := kd.FindBestKernel(ctx)
	if kerr == nil && kernelPath != "" {
		env = append(env,
			"SUPERMIN_KERNEL="+kernelPath,
			"SUPERMIN_MODULES="+modulesDir,
		)
		slog.Debug("Forcing libguestfs appliance build with kernel",
			"kernel", kernelPath,
		)
	} else {
		slog.Warn("No suitable kernel with virtio drivers found in /boot — " +
			"appliance build may hang if the auto-selected kernel lacks virtio")
	}

	// Build environment map for RunCmdCompat (includes current env + kernel overrides)
	runEnv := make(map[string]string, len(env))
	for _, e := range env {
		if key, val, found := strings.Cut(e, "="); found && key != "" {
			runEnv[key] = val
		}
	}

	timeout := 60 * time.Second
	cmdCtx, cmdCancel := context.WithTimeout(ctx, timeout)
	defer cmdCancel()

	result := system.RunCmdCompat(cmdCtx, []string{makeTool, applianceDir}, system.RunCmdOpts{
		Capture: true,
		Check:   true,
		Timeout: timeout,
		Env:     runEnv,
	})
	if result.Err != nil {
		errStr := result.Err.Error()
		if strings.Contains(errStr, "timed out") {
			slog.Warn("libguestfs appliance build timed out after 60s")
			return "", nil
		}
		if strings.Contains(errStr, "Command not found") {
			slog.Warn("libguestfs-make-fixed-appliance command not found")
			return "", nil
		}
		slog.Warn("libguestfs appliance build failed",
			"error", result.Err,
			"output", result.Stdout+result.Stderr,
		)
		return "", nil
	}

	slog.Info("libguestfs fixed appliance built", "path", applianceDir)
	return applianceDir, nil
}

// CleanStaleGuestfsState removes stale libguestfs processes, locks, sockets,
// and caches.
//
// Returns true if any stale state was removed or process was killed.
func (gs *GuestfsService) CleanStaleGuestfsState() bool {
	uid := os.Getuid()
	cleaned := false

	// ── Phase 0: Find and kill abandoned guestfs processes ──────────────
	// Matches Python's ProcessSignalHandler.terminate_batch():
	//   sends SIGTERM to ALL PIDs, waits graceful_timeout=0.5s,
	//   then SIGKILL for survivors.
	abandonedPids := gs.findAbandonedGuestfsProcesses(uid)
	if len(abandonedPids) > 0 {
		// Phase 0a: SIGTERM all PIDs (batch)
		for _, pid := range abandonedPids {
			if proc, err := os.FindProcess(pid); err == nil {
				proc.Signal(syscall.SIGTERM) // ignore error — process may already be gone
			}
		}

		// Phase 0b: Wait for graceful shutdown (matches Python's graceful_timeout=0.5)
		time.Sleep(500 * time.Millisecond)

		// Phase 0c: SIGKILL survivors
		for _, pid := range abandonedPids {
			if proc, err := os.FindProcess(pid); err == nil {
				proc.Kill() // ignore error — may already have exited
			}
		}
		cleaned = true
	}

	// ── Phase 1: Remove the global lock file ────────────────────────────
	lockFile := filepath.Join("/var/tmp", fmt.Sprintf(".guestfs-%d", uid), "lock")
	if _, err := os.Stat(lockFile); err == nil {
		if err := os.Remove(lockFile); err == nil {
			cleaned = true
			slog.Debug("Removed stale libguestfs lock", "path", lockFile)
		}
	}

	// ── Phase 2: Remove stale daemon sockets ────────────────────────────
	sockBase := filepath.Join("/run/user", strconv.Itoa(uid))
	sockDirs, err := filepath.Glob(filepath.Join(sockBase, "libguestfs*"))
	if err == nil {
		for _, sockDir := range sockDirs {
			socks, _ := filepath.Glob(filepath.Join(sockDir, "guestfsd.sock"))
			for _, sock := range socks {
				if err := os.Remove(sock); err == nil {
					cleaned = true
					slog.Debug("Removed stale libguestfs socket", "path", sock)
				}
			}
		}
	}

	// ── Phase 3: Remove cached appliance directories in /var/tmp ────────
	guestfsTmp := filepath.Join("/var/tmp", fmt.Sprintf(".guestfs-%d", uid))
	if entries, err := os.ReadDir(guestfsTmp); err == nil {
		for _, entry := range entries {
			if entry.IsDir() && strings.HasPrefix(entry.Name(), "appliance.d") {
				appliancePath := filepath.Join(guestfsTmp, entry.Name())
				if err := os.RemoveAll(appliancePath); err == nil {
					cleaned = true
					slog.Debug("Removed stale libguestfs cache", "path", appliancePath)
				}
			}
		}
	}

	return cleaned
}

// findAbandonedGuestfsProcesses finds QEMU/guestfish PIDs owned by uid that
// are running the libguestfs appliance but have no mvmctl ancestor.
func (gs *GuestfsService) findAbandonedGuestfsProcesses(uid int) []int {
	var abandoned []int

	procDir := "/proc"
	entries, err := os.ReadDir(procDir)
	if err != nil {
		return abandoned
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		pid, err := strconv.Atoi(entry.Name())
		if err != nil {
			continue
		}

		// Check ownership
		statusPath := filepath.Join(procDir, entry.Name(), "status")
		statusData, err := os.ReadFile(statusPath)
		if err != nil {
			continue
		}
		procUid := system.ParseProcStatusField(string(statusData), "Uid:")
		if procUid != uid {
			continue
		}

		// Read cmdline
		cmdlinePath := filepath.Join(procDir, entry.Name(), "cmdline")
		cmdlineData, err := os.ReadFile(cmdlinePath)
		if err != nil {
			continue
		}
		cmdline := string(cmdlineData)

		// Match QEMU processes running the libguestfs appliance,
		// or guestfish processes left behind
		if strings.Contains(cmdline, ".guestfs-") &&
			(strings.Contains(cmdline, "appliance.d") || strings.Contains(cmdline, "guestfsd.sock")) {
			if !system.HasAncestorWithCmdline(pid, "mvm") {
				abandoned = append(abandoned, pid)
			}
		} else if strings.Contains(strings.ToLower(cmdline), "guestfish") {
			if !system.HasAncestorWithCmdline(pid, "mvm") {
				abandoned = append(abandoned, pid)
			}
		}
	}

	return abandoned
}

// PruneAppliance removes the libguestfs appliance folder and stale system state.
func (gs *GuestfsService) PruneAppliance(cacheDir string, dryRun bool) bool {
	applianceDir := filepath.Join(cacheDir, "appliance")
	removed := false

	if _, err := os.Stat(applianceDir); err == nil {
		if !dryRun {
			if err := os.RemoveAll(applianceDir); err == nil {
				removed = true
			}
		} else {
			removed = true
		}
	}

	if !dryRun {
		stateCleaned := gs.CleanStaleGuestfsState()
		removed = removed || stateCleaned
	}

	return removed
}
